File: app/src/main/java/com/summerland/android/twozerofoureightapp/MainActivity.java
package com.summerland.android.twozerofoureightapp;

import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.MotionEvent;
import android.view.View;
import android.widget.TextView;
import android.widget.Toast;

import java.util.ArrayList;

public class MainActivity extends AppCompatActivity {

    static private float MIN_DISTANCE = 200;
    private float x1 = 0, y1 = 0, x2 = 0, y2 = 0;
    private TwoZeroFourEight game;
    private int[] cells = new int[16];

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        cells = new int[16];
        cells[0] = R.id.index0;
        cells[1] = R.id.index1;
        cells[2] = R.id.index2;
        cells[3] = R.id.index3;
        cells[4] = R.id.index4;
        cells[5] = R.id.index5;
        cells[6] = R.id.index6;
        cells[7] = R.id.index7;
        cells[8] = R.id.index8;
        cells[9] = R.id.index9;
        cells[10] = R.id.index10;
        cells[11] = R.id.index11;
        cells[12] = R.id.index12;
        cells[13] = R.id.index13;
        cells[14] = R.id.index14;
        cells[15] = R.id.index15;

        Log.i("Logm", "Inside on create main activity method");

        if (savedInstanceState != null) {
            TwoZeroFourEight tmpGame = (TwoZeroFourEight) savedInstanceState.getSerializable("GAME_KEY");
            if (tmpGame != null) {
                game = tmpGame;
                game.rePlot();
            }
        } else {
            game = new TwoZeroFourEight();
        }
        paintTiles();
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        Log.i("Logm", "Inside onSaveInstanceState method");
        outState.putSerializable("GAME_KEY", game);
        super.onSaveInstanceState(outState);
    }

    @Override
    protected void onRestoreInstanceState(Bundle savedInstanceState) {
        super.onRestoreInstanceState(savedInstanceState);
        Log.i("Logm", "Inside onRestoreInstanceState method");
        if (savedInstanceState != null) {
            TwoZeroFourEight tmpGame = (TwoZeroFourEight) savedInstanceState.getSerializable("GAME_KEY");
            if (tmpGame != null) {
                game = tmpGame;
                game.rePlot();
            }
        }
        paintTiles();
    }

    public void onClick(View view) {
        Log.i("Logm", "Inside on click main activity action method");
        Intent intent = new Intent(this, MainActivity.class);
        startActivity(intent);
    }

    // onTouchEvent () method gets called when User performs any touch event on screen
    // Method to handle touch event like left to right swap and right to left swap
    public boolean onTouchEvent(MotionEvent event) {

        switch (event.getAction()) {
            // when user first touches the screen we get x and y coordinate
            case MotionEvent.ACTION_DOWN: {
                x1 = event.getX();
                y1 = event.getY();
                break;
            }
            case MotionEvent.ACTION_UP: {
                x2 = event.getX();
                y2 = event.getY();

                //if left to right sweep event on screen
                if (x1 < x2 && x2 - x1 > MIN_DISTANCE) {
                    //Toast.makeText(this, "Left to Right Swap Performed", Toast.LENGTH_SHORT).show();
                    if (game.actionMoveRight()) game.addNewTile();
                    paintTiles();
                }

                // if right to left sweep event on screen
                if (x1 > x2 && x1 - x2 > MIN_DISTANCE) {
                    //Toast.makeText(this, "Right to Left Swap Performed", Toast.LENGTH_SHORT).show();
                    if (game.actionMoveLeft()) game.addNewTile();
                    paintTiles();
                }

                // if UP to Down sweep event on screen
                if (y1 < y2 && y2 - y1 > MIN_DISTANCE) {
                    //Toast.makeText(this, "UP to Down Swap Performed", Toast.LENGTH_SHORT).show();
                    if (game.actionMoveDown()) game.addNewTile();
                    paintTiles();
                }

                //if Down to UP sweep event on screen
                if (y1 > y2 && y1 - y2 > MIN_DISTANCE) {
                    //Toast.makeText(this, "Down to UP Swap Performed", Toast.LENGTH_SHORT).show();
                    if (game.actionMoveUp()) game.addNewTile();
                    paintTiles();
                }
                break;
            }
        }
        return super.onTouchEvent(event);
    }

    private void paintTiles() {

        if (game.getMaxTile() >= TwoZeroFourEight.TARGET) {
            Toast.makeText(this, "YOU HAVE WON: " + game.getMaxTile(), Toast.LENGTH_LONG).show();
        } else if (!game.hasMovesRemaining()) {
            Toast.makeText(this, "SORRY, NO MORE MOVES.  PLAY AGAIN?", Toast.LENGTH_LONG).show();
        }

        Log.i("Logm", game.toString());

        TextView tv;
        tv = (TextView) findViewById(R.id.score);
        tv.setText("Score: " + game.getScore());

        ArrayList<TwoZeroFourEight.Transition> transitions = game.getTransitions();
        if (transitions == null || transitions.size() == 0) return;

        for (TwoZeroFourEight.Transition trans : transitions) {
            paintTransition(trans);
        }

/*        for (int i = 0; i < TwoZeroFourEight.GRID_CNT; i++) {
            if (game.getValue(i) == 0)
                paintCell(findViewById(cells[i]), game.getValue(i));
        }*/

        game.createNewTransitions();

        /*paintCell(findViewById(R.id.index0), game.getValue(0));
        paintCell(findViewById(R.id.index1), game.getValue(1));
        paintCell(findViewById(R.id.index2), game.getValue(2));
        paintCell(findViewById(R.id.index3), game.getValue(3));
        paintCell(findViewById(R.id.index4), game.getValue(4));
        paintCell(findViewById(R.id.index5), game.getValue(5));
        paintCell(findViewById(R.id.index6), game.getValue(6));
        paintCell(findViewById(R.id.index7), game.getValue(7));
        paintCell(findViewById(R.id.index8), game.getValue(8));
        paintCell(findViewById(R.id.index9), game.getValue(9));
        paintCell(findViewById(R.id.index10), game.getValue(10));
        paintCell(findViewById(R.id.index11), game.getValue(11));
        paintCell(findViewById(R.id.index12), game.getValue(12));
        paintCell(findViewById(R.id.index13), game.getValue(13));
        paintCell(findViewById(R.id.index14), game.getValue(14));
        paintCell(findViewById(R.id.index15), game.getValue(15));*/
    }

    private void paintTransition(TwoZeroFourEight.Transition trans) {

        TextView tv = (TextView) findViewById(cells[trans.posFinal]);

        if (trans.type == TwoZeroFourEight.actions.COMPACT) {
            Log.i("Logm", "About to do paint compact");
            int w = tv.getWidth();
            int h = tv.getHeight();
            float t = tv.getTextSize();
            Log.i("Logm", "About to do paint compact w=" + w + " h=" + h + " t=" + t);
            try {
                //tv.setShadowLayer(3, 4, Color.GREEN);
                //tv.setPressed(true);
                //tv.setHeight(h+5);
                //tv.setTextSize(getResources().getDimension(R.dimen.tile_txt_compacted));
                //tv.setShadowLayer(3, 3, 3, getResources().getColor(R.color.t_Merge, null));
                paintCell(tv, trans.val);
                wait(500);
                //tv.setRotation(2);
            } catch (Exception e) {
                e.printStackTrace();
            } finally {
                paintCell(tv, trans.val);
            }
            //tv.setWidth(w);
            //tv.setHeight(h);
            //tv.setTextSize(getResources().getDimension(R.dimen.tile_txt_normal));
            //tv.setPressed(false);
        } else {
            paintCell(tv, trans.val);
        }
    }

    private void paintCell(Object obj, int val) {

        TextView tv = (TextView) obj;

        if (val == 0) {
            tv.setText("");
        } else {
            tv.setText("" + val);
        }

        int tx_col = getResources().getColor(R.color.t_dark_text, null);
        int bg_col = getResources().getColor(R.color.t0_bg, null);

        switch (val) {
            case 2:
                bg_col = getResources().getColor(R.color.t2_bg, null);
                break;
            case 4:
                bg_col = getResources().getColor(R.color.t4_bg, null);
                break;
            case 8:
                tx_col = getResources().getColor(R.color.t_light_text, null);
                bg_col = getResources().getColor(R.color.t8_bg, null);
                break;
            case 16:
                tx_col = getResources().getColor(R.color.t_light_text, null);
                bg_col = getResources().getColor(R.color.t16_bg, null);
                break;
            case 32:
                tx_col = getResources().getColor(R.color.t_light_text, null);
                bg_col = getResources().getColor(R.color.t32_bg, null);
                break;
            case 64:
                tx_col = getResources().getColor(R.color.t_light_text, null);
                bg_col = getResources().getColor(R.color.t64_bg, null);
                break;
            case 128:
                tx_col = getResources().getColor(R.color.t_light_text, null);
                bg_col = getResources().getColor(R.color.t128_bg, null);
                break;
            case 256:
                bg_col = getResources().getColor(R.color.t256_bg, null);
                break;
            case 512:
                bg_col = getResources().getColor(R.color.t512_bg, null);
                break;
            case 1024:
                bg_col = getResources().getColor(R.color.t1024_bg, null);
                break;
            case 2048:
                bg_col = getResources().getColor(R.color.t2048_bg, null);
                break;
            default:
                break;
        }
        tv.setBackgroundColor(bg_col);
        tv.setTextColor(tx_col);
    }
}

File: app/src/main/java/com/summerland/android/twozerofoureightapp/MainActivity.kt
package com.summerland.android.twozerofoureightapp

import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.MotionEvent
import android.view.View
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.navigation.ui.AppBarConfiguration
import com.summerland.android.twozerofoureightapp.databinding.ActivityMainBinding
import com.summerland.android.tzfe.game.engine.*

class MainActivity : AppCompatActivity(), GameEngineProtocol {

    private lateinit var appBarConfiguration: AppBarConfiguration
    private lateinit var binding: ActivityMainBinding

    private var x1 = 0f
    private var y1 = 0f

    private lateinit var game: TwoZeroFourEight
    private lateinit var dataStore : StoredDataUtils
    private var highScore : Int = 0
    private val cells = IntArray(16)


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        setSupportActionBar(binding.toolbar)

        setContentView(R.layout.activity_main)
        dataStore = StoredDataUtils(this.baseContext)

        Log.i("Logm", "Inside on create main activity method")

        cells[0] = R.id.index0
        cells[1] = R.id.index1
        cells[2] = R.id.index2
        cells[3] = R.id.index3
        cells[4] = R.id.index4
        cells[5] = R.id.index5
        cells[6] = R.id.index6
        cells[7] = R.id.index7
        cells[8] = R.id.index8
        cells[9] = R.id.index9
        cells[10] = R.id.index10
        cells[11] = R.id.index11
        cells[12] = R.id.index12
        cells[13] = R.id.index13
        cells[14] = R.id.index14
        cells[15] = R.id.index15

        if (savedInstanceState != null) {
            val tmpGame : TwoZeroFourEight? = savedInstanceState.getSerializable(GAME_KEY) as TwoZeroFourEight?
            if (tmpGame != null) {
                game = tmpGame
                game.replotBoard()
            }
        } else {
            game = TwoZeroFourEight(this)
            this.setupNewGame()
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        Log.i("Logm", "Inside onSaveInstanceState method")
        outState.putSerializable(GAME_KEY, game)
        super.onSaveInstanceState(outState)
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        super.onRestoreInstanceState(savedInstanceState)
        Log.i("Logm", "Inside onRestoreInstanceState method")
        val tmpGame :TwoZeroFourEight? = savedInstanceState.getSerializable(GAME_KEY) as TwoZeroFourEight?
        if (tmpGame != null) {
            game = tmpGame
            game.replotBoard()
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu; this adds items to the action bar if it is present.
//        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        return when (item.itemId) {
//            R.id.action_settings -> true
            else -> super.onOptionsItemSelected(item)
        }
    }

//    override fun onSupportNavigateUp(): Boolean {
//        val navController = findNavController(R.id.nav_host_fragment_content_main)
//        return navController.navigateUp(appBarConfiguration)
//                || super.onSupportNavigateUp()
//    }

    @Suppress("UNUSED_PARAMETER")
    fun onClick(view: View) {
        Log.i("Logm", "Inside on click main activity action method")
        this.setupNewGame()
    }

    // Start / Reset game here ->
    private fun setupNewGame() {
        this.highScore = dataStore.getHighscore()
        //this.gamePanel.resetBoard()
        game.newGame(this.highScore)
        this.userScoreChanged(0)
    }

    // onTouchEvent () method gets called when User performs any touch event on screen
    // Method to handle touch event like left to right swap and right to left swap
    override fun onTouchEvent(event: MotionEvent): Boolean {

        when (event.action) {
            // when user first touches the screen we get x and y coordinate
            MotionEvent.ACTION_DOWN -> {
                x1 = event.x
                y1 = event.y
            }
            MotionEvent.ACTION_UP -> {
                val x2 = event.x
                val y2 = event.y
                val minDistance = 200

                if (x1 < x2 && x2 - x1 > minDistance) { game.actionMove(GameMoves.Right) }
                else if (x1 > x2 && x1 - x2 > minDistance) { game.actionMove(GameMoves.Left) }
                else if (y1 < y2 && y2 - y1 > minDistance) { game.actionMove(GameMoves.Down) }
                else if (y1 > y2 && y1 - y2 > minDistance) { game.actionMove(GameMoves.Up) }
            }
        }
        return super.onTouchEvent(event)
    }

    private fun paintTransition(move: GameEngine.Transition) {

        val tv = findViewById<TextView>(cells[move.location])

        if (move.action == TileMoveType.Slide || move.action == TileMoveType.Merge) {
            //Log.i("Logm", "About to do paint compact")
            //Log.i("Logm", "About to do paint compact w=$w h=$h t=$t")
            paintCell(tv, move.value)
            this.paintTransition(GameEngine.Transition(TileMoveType.Clear, 0, move.oldLocation))
        } else {
            paintCell(tv, move.value)
        }
    }

    private fun paintCell(obj: Any, value: Int) {

        val tv = obj as TextView
        tv.text = if (value <= 0) "" else "$value"

        var txCol = resources.getColor(R.color.t_dark_text, null)
        var bgCol = resources.getColor(R.color.t0_bg, null)

        when (value) {
            2 -> bgCol = resources.getColor(R.color.t2_bg, null)
            4 -> bgCol = resources.getColor(R.color.t4_bg, null)
            8 -> {
                txCol = resources.getColor(R.color.t_light_text, null)
                bgCol = resources.getColor(R.color.t8_bg, null)
            }
            16 -> {
                txCol = resources.getColor(R.color.t_light_text, null)
                bgCol = resources.getColor(R.color.t16_bg, null)
            }
            32 -> {
                txCol = resources.getColor(R.color.t_light_text, null)
                bgCol = resources.getColor(R.color.t32_bg, null)
            }
            64 -> {
                txCol = resources.getColor(R.color.t_light_text, null)
                bgCol = resources.getColor(R.color.t64_bg, null)
            }
            128 -> {
                txCol = resources.getColor(R.color.t_light_text, null)
                bgCol = resources.getColor(R.color.t128_bg, null)
            }
            256 -> bgCol = resources.getColor(R.color.t256_bg, null)
            512 -> bgCol = resources.getColor(R.color.t512_bg, null)
            1024 -> bgCol = resources.getColor(R.color.t1024_bg, null)
            2048 -> bgCol = resources.getColor(R.color.t2048_bg, null)
            else -> { } // won't happen
        }
        tv.setBackgroundColor(bgCol)
        tv.setTextColor(txCol)
    }

    override fun userWin() {
        Toast.makeText(this, resources.getString(R.string.winner_toast_message) +
                game.maxTile, Toast.LENGTH_LONG).show()
    }

    override fun userFail() {
        Toast.makeText(this, resources.getString(R.string.lose_toast_message),
            Toast.LENGTH_LONG).show()
    }

    override fun userPB(score: Int) {
        Toast.makeText(this, resources.getString(R.string.personalbest_toast_message),
            Toast.LENGTH_SHORT).show()
        this.dataStore.putHighscore(score)
        this.highScore = score
    }

    override fun userScoreChanged(score: Int) {
        val tv: TextView = findViewById(R.id.score)
        tv.text = StringBuffer(resources.getString(R.string.score)).append(score)
    }

    override fun updateTileValue(move: GameEngine.Transition) {
        paintTransition(move)
    }

    companion object {
        private const val GAME_KEY = "2048_GAME_KEY"
    }

}